### DISCUSSION TITLE
Game controler : Problème d'installation

### DISCUSSION THREAD

**Discussion by RVDROU** [robot-soccer-kit/robot-soccer-kit#41] [Q&A]
Bonjour,
Tout d'abord un grand merci pour ce projet, la qualité de l'ensemble est remarquable.
J'ai commencé à installer l'ensemble mais le démarrage du contrôleur de jeu renvoie l'erreur  'cv2 has no attribut aruco'. Avez vous déjà rencontré ce problème ?
Voici les actions que j'ai effectué et le matéreil utilisé :
Matériel :
    - PC fixe I3 avec 8Go de ram / DD: 80Go
    - OS : Linux MINT 20.3 (dernière version)
Installation :
   - Python 3 + PIP + robot-soccer-kit comme indiqué sur la doc
   - La réinstallation de la bibliothèque indique que toutes les bibliothèques sont installées

Je dois aussi avoir un problème avec le bluetooth, je posterai un autre message après avoir résolu ce premier souci.
Merci d'avance pour votre aide

Hervé D (Périgueux)

**Comment by Gregwar** [comment]
Est-ce que vous avez installé le paquet avec les dépendances du Game Controller ?

```
pin install robot-soccer-kit[gc]
```

Ici le `[gc]` indique qu'on installe aussi les dépendances du game controller (cf https://github.com/robot-soccer-kit/robot-soccer-kit/blob/master/setup.py#L35) qui sont plus lourdes que d'installer simplement le client

**Reply by RVDROU** [reply]
Oui j'ai bien installé le paquet pour le serveur ([robot-soccer-kit[gc]) avec pip.

**Reply by Gregwar** [reply]
C'est quelle version de python ?
Quel est le résultat de la commande `pip freeze` ?

**Reply by cgaspard3333** [reply]
Sinon peut être un problème de version d'OpenCV.
Essayez de : 

1. Désinstaller OpenCV avec la commande : `pip uninstall opencv-python` puis `pip uninstall opencv-contrib-python`

1. Réinstaller avec la commande : `pip install opencv-contrib-python`

**Reply by RVDROU** [reply]
Merci pour vos réponses, je testerai tout ça demain au lycée. Je vous tiendrai informés des résultats.
Bonne soirée

**Reply by RVDROU** [reply]
J'ai suivi vos remarques sans succès pour le moment. Voilà les infos demandées :
   - Version de Python : 3.8.10 (voir rapport joint)
   - Les bibliothèques [gc] sont référencées dans pip freeze (voir rapport joint)
   - La désinstallation puis réinstall de opencv-python et opencv-contrib-python a été faîte
Je ne sais pas trop comment résoudre le pb j'espère que le rapport joint pourra vous aider.

Merci d'avance
[rapport_bug.txt](https://github.com/robot-soccer-kit/robot-soccer-kit/files/8215437/rapport_bug.txt)

**Reply by cgaspard3333** [reply]
Il faudrait essayer de désinstaller les packages suivants : 

1. opencv-contrib-python-headless (`pip3 uninstall opencv-contrib-python-headless` / `pip3 uninstall opencv-contrib-python` )
2. opencv-python (`pip3 uninstall opencv-python`)
3. opencv-python-headless (`pip3 uninstall opencv-python-headless` / `pip3 uninstall opencv-python`)
4. robot-soccer-kit (`pip3 uninstall robot-soccer-kit`)

**Ne pas réinstaller le package opencv-python !!**

Installer OpenCV avec la commande `pip3 install opencv-contrib-python`
Installer le package RSK avec Game Controller : `pip3 install -U robot-soccer-kit[gc]`

**Reply by RVDROU** [reply]
Rien n'y fait, j'ai donc décidé de tout reprendre à zéro depuis un PC vierge :
  - Installation de LINUX MINT 20.3 (dernière version)
  - Mise à jour système
  - Installation du controleur :
            -> sudo apt-get update
            -> sudo apt-get install python3  (déjà présent)
            -> sudo apt-get install python3-pip
            -> sudo pip install -U robot-soccer-kit[gc]
L'exécution de la commande >>> python3 -m rsk.game_controller  renvoie la même erreur ([rapport_bug2.txt](https://github.com/robot-soccer-kit/robot-soccer-kit/files/8218168/rapport_bug2.txt)). 

Avez vous testé l'installation sur Linux et si oui sur quelle distribution ?  Je ne pense pas que la distribution puisse avoir une quelconque importance, mais je ne vois pas d'autre piste.

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[rapport_bug2.txt](https://github.com/robot-soccer-kit/robot-soccer-kit/files/8218165/rapport_bug2.txt)

**Reply by cgaspard3333** [reply]
Demain dans la journée ou vendredi je suis disponible pour faire un appel avec prise de contrôle à distance avec TeamViewer par exemple pour essayer de voir d'où pourrait venir le problème. 
Vous pouvez me contacter directement par mail pour qu'on échange nos coordonnées et pour organiser cela 
Je vais essayer de réaliser l'installation de Linux Mint d'ici là dans une VM pour localiser le problème. Et noter toutes les étapes d'installation que je réalise.

**Reply by cgaspard3333** [reply]
Je viens de finir d'installer la machine virtuelle avec Linux Mint 20.3 et de tout tester. 
J'ai noté les seules commandes que j'ai tapé dans un terminal après le démarrage de la machine et dans cet ordre : 
1. `sudo apt update`
2. `sudo apt upgrade`
3. `sudo apt install python3-pip`
4. `pip3 install -U robot-soccer-kit[gc]`
5. `python3 -m rsk.game_controller`

Lors de la 5ème commande pour l'ouverture du Game Controller, il a démarré directement, sans soucis et sans aucun message d'erreur.
Je me dis que l'erreur peut peut-être venir de `pip3` au lieu de `pip` à la commande 4.
Essayez de réaliser ça sur une machine avec de nouveau une installation propre, il ne devrait pas y avoir de raison que cela ne fonctionne pas.

**Reply by RvClaveille** [reply]
Toujours pareil, c'est un truc de dingue. J'ai suivi exactement ta démarche -> Installation propre + Mise à jour + installation PIP + Installation RSK [gc].
Je joins les retours de chaque commande. A noter que l'installation de RSK me renvoie des WARNING (problème de PATH), est ce que le problème vient de là ?
Sinon le message est renvoyé par le script de détection du terrain, est ce que ça peut venir de la camera ??
[rapport_bug3.txt](https://github.com/robot-soccer-kit/robot-soccer-kit/files/8222772/rapport_bug3.txt)
?

**Reply by fred-hautbrion** [reply]
nous on a une erreur :
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-ff51qw2c/opencv-contrib-python-headless/
et en installant opencv on a : 
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-iz6ukms7/opencv-contrib-python/
et on est bloqué là... pour l'installation du game controleur (le client s'est installé sans soucis)
notre distrib : xubuntu 18.04
Fred

**Reply by cgaspard3333** [reply]
Pour @RVDROU normalement ça fonctionne après appel et TeamViewer

Maintenant pour @fred-hautbrion 
Tout d'abord : 

1. Vérifiez que Python est à jour et avec uniquement une version 3.8 ou 3.9.
2. Verifiez que pip et ubuntu sont aussi à jour.

Je n'ai jamais eu cette erreur jusqu'à présent...

Essayez aussi cette technique de mon post précédent : 

> Il faudrait essayer de désinstaller les packages suivants :
> 
> 1. opencv-contrib-python-headless (`pip3 uninstall opencv-contrib-python-headless` / `pip3 uninstall opencv-contrib-python` )
> 2. opencv-python (`pip3 uninstall opencv-python`)
> 3. opencv-python-headless (`pip3 uninstall opencv-python-headless` / `pip3 uninstall opencv-python`)
> 4. robot-soccer-kit (`pip3 uninstall robot-soccer-kit`)
> 
> **Ne pas réinstaller le package opencv-python !!**
> 
> Installer OpenCV avec la commande `pip3 install opencv-contrib-python` Installer le package RSK avec Game Controller : `pip3 install -U robot-soccer-kit[gc]`

Sinon, il faudrait envisager de mettre à jour le kernel de Linux, voici la procédure : 

```
sudo add-apt-repository ppa:cappelikan/ppa
sudo apt update
sudo apt install mainline
```

Le logiciel sera disponible dans le menu Système → Ubuntu Mainline kernel installer ou en console par la commande `mainline`

Lorsqu'il s'ouvre, selectionnez la dernière version du kernel disponible puis "install"

Redémarrez à la fin de la mise à jour et essayer de réinstaller avec la commande `pip3 install -U robot-soccer-kit[gc]`

**Reply by fred-hautbrion** [reply]
ça y est, je suis arrivé à installer le robot-soccer-kit[gc] après un update de ubuntu, la mise en place de python3.8 et un update de pip en mode user : pip install --upgrade pip --user
cependant lorsque je lance python -m rsk.game_controller voici le message 

/usr/bin/python3.8: Error while finding module specification for 'rsk.game_controller' (ModuleNotFoundError: No module named 'rsk')

python --version me donne bien python3.8.0

Je n'ai pas encore installer le bluetooth, c'est peut être ça?

**Reply by cgaspard3333** [reply]
De prime abord, j'aurais l'impression que cela vient du fait qu'il y aurait plusieurs installations de Python sur le même PC. Par exemple, que RSK aurait été installé sur une version "admin" de Python et que vous essayez de le lancer depuis une version "user" de Python. Ou alors tout simplement deux installations de Python en parallèle. 
Ce qui expliquerait que Python ne trouve pas le package RSK d'installé.

Je suis dispo par téléphone pour faire un TeamViewer aujourd'hui si jamais vous êtes disponible.

**Reply by fred-hautbrion** [reply]
J'ai enlevé tous les pythons sauf le 3.8 mais même constat... Je vais passer en xubuntu 20.4...

**Reply by RVDROU** [reply]
Mon installation avec controleur sous MINT 20.3 fonctionne. Après quelques soucis pour détecter les coins du terrain, tout est installé et opérationnel. Le problème venait des markers ques j'avais découpé sans laisser de bord blanc sur l'extérieur.
Encore un grand merci à l'équipe pour les conseils et la hot line, mes élèves sont ravis du résultat et moi aussi.
A bientôt

**Reply by fred-hautbrion** [reply]
J'ai résolue mon souci : pip se lançait avec python3.6 malgré la désistallation de ce dernier (là je n'ai pas compris). Bref en installant pip avec pip lui même, je fonctionne sous pip version python3.8. La ligne de code salvatrice : 
`python -m pip install pip`
Puis après reprendre les installations `pip3 install -U robot-soccer-kit[gc]`
